File: azkar/apple/iOS/App/Reminders.swift
// The reminders iOS delivers while Azkar isn't running: one notification per planned card.
import Foundation
import UserNotifications

enum Reminders {
  /// iOS keeps at most 64 pending notifications per app, which is how long a plan is.
  static let limit = 64

  private static var center: UNUserNotificationCenter { .current() }

  /// Asks the first time; after that it only reports the answer given.
  static func authorize() async -> UNAuthorizationStatus {
    let status = await center.notificationSettings().authorizationStatus
    guard status == .notDetermined else { return status }
    _ = try? await center.requestAuthorization(options: [.alert, .sound])
    return await center.notificationSettings().authorizationStatus
  }

  /// Replaces everything still pending with this plan. Delivered notifications are left alone:
  /// those are the azkar waiting to be opened.
  static func schedule(_ plan: [PlannedReminder], sound: Bool) async {
    center.removeAllPendingNotificationRequests()
    let calendar = Calendar.current
    for (i, reminder) in plan.enumerated() {
      let content = UNMutableNotificationContent()
      content.title = cardHeader(reminder.card)
      content.body = notificationBody(reminder.card.zikr, last: i == plan.count - 1)
      // One thread per list, so the morning azkar stack together and the evening ones apart.
      content.threadIdentifier = reminder.card.session.rawValue
      content.sound = sound ? .default : nil
      // The card the notification opens travels with it, so a tap needs no guesswork.
      content.userInfo = ["azkar": PlanFile.encode([reminder])]
      let parts = calendar.dateComponents([.year, .month, .day, .hour, .minute, .second], from: reminder.fireAt)
      let request = UNNotificationRequest(
        identifier: "azkar-\(i)",
        content: content,
        trigger: UNCalendarNotificationTrigger(dateMatching: parts, repeats: false))
      try? await center.add(request)
    }
  }

  /// How many reminders are still to come.
  static func pending() async -> Int { await center.pendingNotificationRequests().count }
}

File: azkar/kotlin/shared/src/iosMain/kotlin/com/wildduck/azkar/ios/Reminders.kt
// The reminders iOS delivers while Azkar isn't running: one notification per planned card.
package com.wildduck.azkar.ios

import com.wildduck.azkar.app.PlanFile
import com.wildduck.azkar.app.cardHeader
import com.wildduck.azkar.app.notificationBody
import com.wildduck.azkar.core.Card
import com.wildduck.azkar.core.PlannedReminder
import kotlinx.datetime.TimeZone
import kotlinx.datetime.toLocalDateTime
import platform.Foundation.NSDateComponents
import platform.UserNotifications.UNAuthorizationOptionAlert
import platform.UserNotifications.UNAuthorizationOptionSound
import platform.UserNotifications.UNCalendarNotificationTrigger
import platform.UserNotifications.UNMutableNotificationContent
import platform.UserNotifications.UNNotificationRequest
import platform.UserNotifications.UNNotificationSound
import platform.UserNotifications.UNUserNotificationCenter

/** The key a notification carries its own reminder under, so a tap needs no guesswork. */
private const val KEY = "azkar"

object Reminders {
    /** iOS keeps at most 64 pending notifications per app, which is how long a plan is. */
    const val LIMIT = 64

    private val center get() = UNUserNotificationCenter.currentNotificationCenter()

    /** Asks the first time; after that iOS answers with what was said then. */
    fun authorize(onAnswer: (Boolean) -> Unit) {
        center.requestAuthorizationWithOptions(UNAuthorizationOptionAlert or UNAuthorizationOptionSound) {
            granted, _ ->
            onAnswer(granted)
        }
    }

    /**
     * Replaces everything still pending with this plan. Delivered notifications are left alone: those are the
     * azkar waiting to be opened.
     */
    fun schedule(
        plan: List<PlannedReminder>,
        sound: Boolean,
        timeZone: TimeZone = TimeZone.currentSystemDefault(),
    ) {
        center.removeAllPendingNotificationRequests()
        plan.forEachIndexed { i, reminder ->
            val content = UNMutableNotificationContent().apply {
                setTitle(cardHeader(reminder.card))
                setBody(notificationBody(reminder.card.zikr, last = i == plan.size - 1))
                // One thread per list, so the morning azkar stack together and the evening ones apart.
                setThreadIdentifier(reminder.card.session.name.lowercase())
                if (sound) setSound(UNNotificationSound.defaultSound)
                setUserInfo(mapOf(KEY to PlanFile.encode(listOf(reminder))))
            }
            val at = reminder.fireAt.toLocalDateTime(timeZone)
            val parts = NSDateComponents().apply {
                year = at.year.toLong()
                month = (at.month.ordinal + 1).toLong()
                day = at.day.toLong()
                hour = at.hour.toLong()
                minute = at.minute.toLong()
                second = at.second.toLong()
            }
            center.addNotificationRequest(
                UNNotificationRequest.requestWithIdentifier(
                    identifier = "azkar-$i",
                    content = content,
                    trigger = UNCalendarNotificationTrigger.triggerWithDateMatchingComponents(parts, false),
                ),
                null,
            )
        }
    }

    /** The card a delivered notification was made from. */
    fun card(userInfo: Map<Any?, *>?): Card? {
        val text = userInfo?.get(KEY) as? String ?: return null
        return PlanFile.decode(text).firstOrNull()?.card
    }
}
